# Project -  Data Preparation I

This is the second part of the AC data mining project. In this notebook, you will find the proccess of data preparation consisting in merging some of the tables, access data quality, outlier treatment and feature engineering.

## Import of the needed packages

In [4]:
import pandas as pd 

In [27]:
#Carregando os valores do dataset
base = pd.read_csv('quinto_andar_transformed_data.csv')
basecpy= pd.read_csv('quinto_andar_transformed_data.csv')

#Mostrando as 5 primeiras linhas do dataset
base.head()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2775 entries, 0 to 2774
Data columns (total 16 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   url              2775 non-null   object 
 1   bairro           2775 non-null   object 
 2   aluguel          2775 non-null   float64
 3   condominio       2768 non-null   float64
 4   iptu             2767 non-null   float64
 5   seguro_incendio  2775 non-null   float64
 6   taxa_serviço     2775 non-null   float64
 7   total            2775 non-null   float64
 8   metragem         2775 non-null   float64
 9   quarto           2775 non-null   float64
 10  banheiro         2775 non-null   float64
 11  vaga_carro       2775 non-null   float64
 12  andar            2775 non-null   float64
 13  aceita_pet       2775 non-null   float64
 14  mobilia          2775 non-null   float64
 15  metro_prox       2775 non-null   float64
dtypes: float64(14), object(2)
memory usage: 347.0+ KB


<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>url</th>
      <th>bairro</th>
      <th>aluguel</th>
      <th>condominio</th>
      <th>iptu</th>
      <th>seguro_incendio</th>
      <th>taxa_serviço</th>
      <th>total</th>
      <th>metragem</th>
      <th>quarto</th>
      <th>banheiro</th>
      <th>vaga_carro</th>
      <th>andar</th>
      <th>aceita_pet</th>
      <th>mobilia</th>
      <th>metro_prox</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>https://www.quintoandar.com.br/imovel/89332996...</td>
      <td>Liberdade</td>
      <td>1900.0</td>
      <td>424.0</td>
      <td>5.0</td>
      <td>25.0</td>
      <td>38.0</td>
      <td>2392.0</td>
      <td>57.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>18.0</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>https://www.quintoandar.com.br/imovel/89334069...</td>
      <td>Santa Cecília</td>
      <td>1190.0</td>
      <td>460.0</td>
      <td>0.0</td>
      <td>16.0</td>
      <td>23.0</td>
      <td>1689.0</td>
      <td>56.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>11.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>https://www.quintoandar.com.br/imovel/89333859...</td>
      <td>Aclimação</td>
      <td>2330.0</td>
      <td>450.0</td>
      <td>5.0</td>
      <td>30.0</td>
      <td>46.0</td>
      <td>2861.0</td>
      <td>48.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>10.0</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>3</th>
      <td>https://www.quintoandar.com.br/imovel/89337065...</td>
      <td>Liberdade</td>
      <td>1850.0</td>
      <td>418.0</td>
      <td>6.0</td>
      <td>24.0</td>
      <td>37.0</td>
      <td>2335.0</td>
      <td>44.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>3.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>4</th>
      <td>https://www.quintoandar.com.br/imovel/89337700...</td>
      <td>Vila Guilherme</td>
      <td>2200.0</td>
      <td>392.0</td>
      <td>10.0</td>
      <td>28.0</td>
      <td>44.0</td>
      <td>2674.0</td>
      <td>62.0</td>
      <td>2.0</td>
      <td>2.0</td>
      <td>1.0</td>
      <td>8.0</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
  </tbody>
</table>
</div>

In [28]:
#Verificando as informações do dataset
base.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2775 entries, 0 to 2774
Data columns (total 16 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   url              2775 non-null   object 
 1   bairro           2775 non-null   object 
 2   aluguel          2775 non-null   float64
 3   condominio       2768 non-null   float64
 4   iptu             2767 non-null   float64
 5   seguro_incendio  2775 non-null   float64
 6   taxa_serviço     2775 non-null   float64
 7   total            2775 non-null   float64
 8   metragem         2775 non-null   float64
 9   quarto           2775 non-null   float64
 10  banheiro         2775 non-null   float64
 11  vaga_carro       2775 non-null   float64
 12  andar            2775 non-null   float64
 13  aceita_pet       2775 non-null   float64
 14  mobilia          2775 non-null   float64
 15  metro_prox       2775 non-null   float64
dtypes: float64(14), object(2)
memory usage: 347.0+ KB


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2775 entries, 0 to 2774
Data columns (total 16 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   url              2775 non-null   object 
 1   bairro           2775 non-null   object 
 2   aluguel          2775 non-null   float64
 3   condominio       2768 non-null   float64
 4   iptu             2767 non-null   float64
 5   seguro_incendio  2775 non-null   float64
 6   taxa_serviço     2775 non-null   float64
 7   total            2775 non-null   float64
 8   metragem         2775 non-null   float64
 9   quarto           2775 non-null   float64
 10  banheiro         2775 non-null   float64
 11  vaga_carro       2775 non-null   float64
 12  andar            2775 non-null   float64
 13  aceita_pet       2775 non-null   float64
 14  mobilia          2775 non-null   float64
 15  metro_prox       2775 non-null   float64
dtypes: float64(14), object(2)
memory usage: 347.0+ KB


É possivel verificar que temos 16 colunas e 2775 linhas no total. Porém, nas colunas condominio e iptu temos menos linhas. Vamos verificar o que se passa 

In [29]:
#Verificando a quantidade de valores nulos  por coluna
base.isnull().sum()

url                0
bairro             0
aluguel            0
condominio         7
iptu               8
seguro_incendio    0
taxa_serviço       0
total              0
metragem           0
quarto             0
banheiro           0
vaga_carro         0
andar              0
aceita_pet         0
mobilia            0
metro_prox         0
dtype: int64

É possível verificar que na coluna condominio e iptu temos 7 e 8 valores nulos respectivamente

url                0
bairro             0
aluguel            0
condominio         7
iptu               8
seguro_incendio    0
taxa_serviço       0
total              0
metragem           0
quarto             0
banheiro           0
vaga_carro         0
andar              0
aceita_pet         0
mobilia            0
metro_prox         0
dtype: int64

In [33]:
base.url[13]





'https://www.quintoandar.com.br/imovel/893377120?house_tags=newAd&search_rank=%7B%22sortMode%22%3A%22relevance%22%2C%22searchMode%22%3A%22list%22%2C%22resultsOrigin%22%3A%22search%22%2C%22rank%22%3A553%2C%22personalization%22%3Atrue%7D&search_id=%224bf69054f19a11eb87180a0adc3b16b0%22&from_route=%22search_results%22'

Entendemos que valores como condómino nulo são registros de casas que realmente não possuem condomínio .                           
Uma casa, especialmente se for uma casa independente e não parte de um condomínio fechado ou comunidade planejada geralmente não paga taxa de condomínio. 

Assim, podemos substituir por 0 esses valores NaN

In [34]:
#filtrando os valores nulos da coluna condominio    

base[base['condominio'].isnull()]

url      bairro  aluguel  \
987   https://www.quintoandar.com.br/imovel/89288689...  Bela Vista   2960.0   
1391  https://www.quintoandar.com.br/imovel/89310205...    Perdizes   5200.0   
1527  https://www.quintoandar.com.br/imovel/89310896...   Aclimação   3000.0   
1801  https://www.quintoandar.com.br/imovel/89284189...    Ipiranga   2600.0   
2287  https://www.quintoandar.com.br/imovel/89288795...    Ipiranga   1000.0   
2387  https://www.quintoandar.com.br/imovel/89288796...    Ipiranga   1375.0   
2751  https://www.quintoandar.com.br/imovel/89288796...    Ipiranga   1375.0   

      condominio    iptu  seguro_incendio  taxa_serviço   total  metragem  \
987          NaN   125.0             49.0          59.0  3193.0     100.0   
1391         NaN  1030.0             86.0         104.0  6420.0     220.0   
1527         NaN   158.0             50.0          60.0  3268.0      90.0   
1801         NaN   100.0             43.0          52.0  2795.0     100.0   
2287         NaN    80.0             17.0          20.0  1117.0      25.0   
2387         NaN   100.0             23.0          27.0  1525.0      50.0   
2751         NaN   100.0             23.0          27.0  1525.0      50.0   

      quarto  banheiro  vaga_carro  andar  aceita_pet  mobilia  metro_prox  
987      2.0       1.0         0.0    0.0         0.0      0.0         1.0  
1391     3.0       3.0         1.0    0.0         1.0      0.0         0.0  
1527     3.0       1.0         1.0    0.0         1.0      0.0         1.0  
1801     2.0       2.0         1.0    0.0         1.0      0.0         0.0  
2287     1.0       1.0         0.0    2.0         0.0      0.0         1.0  
2387     1.0       1.0         0.0    0.0         0.0      0.0         1.0  
2751     1.0       1.0         0.0    0.0         0.0      0.0         1.0

Valores nulos da coluna condominio  
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>url</th>
      <th>bairro</th>
      <th>aluguel</th>
      <th>condominio</th>
      <th>iptu</th>
      <th>seguro_incendio</th>
      <th>taxa_serviço</th>
      <th>total</th>
      <th>metragem</th>
      <th>quarto</th>
      <th>banheiro</th>
      <th>vaga_carro</th>
      <th>andar</th>
      <th>aceita_pet</th>
      <th>mobilia</th>
      <th>metro_prox</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>987</th>
      <td>https://www.quintoandar.com.br/imovel/89288689...</td>
      <td>Bela Vista</td>
      <td>2960.0</td>
      <td>NaN</td>
      <td>125.0</td>
      <td>49.0</td>
      <td>59.0</td>
      <td>3193.0</td>
      <td>100.0</td>
      <td>2.0</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>1391</th>
      <td>https://www.quintoandar.com.br/imovel/89310205...</td>
      <td>Perdizes</td>
      <td>5200.0</td>
      <td>NaN</td>
      <td>1030.0</td>
      <td>86.0</td>
      <td>104.0</td>
      <td>6420.0</td>
      <td>220.0</td>
      <td>3.0</td>
      <td>3.0</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>1527</th>
      <td>https://www.quintoandar.com.br/imovel/89310896...</td>
      <td>Aclimação</td>
      <td>3000.0</td>
      <td>NaN</td>
      <td>158.0</td>
      <td>50.0</td>
      <td>60.0</td>
      <td>3268.0</td>
      <td>90.0</td>
      <td>3.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>1801</th>
      <td>https://www.quintoandar.com.br/imovel/89284189...</td>
      <td>Ipiranga</td>
      <td>2600.0</td>
      <td>NaN</td>
      <td>100.0</td>
      <td>43.0</td>
      <td>52.0</td>
      <td>2795.0</td>
      <td>100.0</td>
      <td>2.0</td>
      <td>2.0</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>2287</th>
      <td>https://www.quintoandar.com.br/imovel/89288795...</td>
      <td>Ipiranga</td>
      <td>1000.0</td>
      <td>NaN</td>
      <td>80.0</td>
      <td>17.0</td>
      <td>20.0</td>
      <td>1117.0</td>
      <td>25.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>2.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>2387</th>
      <td>https://www.quintoandar.com.br/imovel/89288796...</td>
      <td>Ipiranga</td>
      <td>1375.0</td>
      <td>NaN</td>
      <td>100.0</td>
      <td>23.0</td>
      <td>27.0</td>
      <td>1525.0</td>
      <td>50.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>2751</th>
      <td>https://www.quintoandar.com.br/imovel/89288796...</td>
      <td>Ipiranga</td>
      <td>1375.0</td>
      <td>NaN</td>
      <td>100.0</td>
      <td>23.0</td>
      <td>27.0</td>
      <td>1525.0</td>
      <td>50.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.0</td>
    </tr>
  </tbody>
</table>
</div>


In [41]:
#filtrando apenas os valores nulos da coluna condominio 

#base[base['condominio'].isnull()]['condominio']
base.loc[base['condominio'].isnull(), 'condominio']


987    NaN
1391   NaN
1527   NaN
1801   NaN
2287   NaN
2387   NaN
2751   NaN
Name: condominio, dtype: float64